lxd/storage/backend_lxd.go

```diff
@@ -2223,6 +2299,174 @@ func (b *lxdBackend) CreateInstanceFromMigration(inst instance.Instance, conn io
+	if isRemoteClusterMove {
+		return fmt.Errorf("Conversion cannot be used for moving instances between members")
+	}
+
+	contentType := InstanceContentType(inst)
+	volType, err := InstanceTypeToVolumeType(inst.Type())
+	if err != nil {
+		return err
+	}
+
+	volStorageName := project.Instance(inst.Project().Name, inst.Name())
+	volConfig := make(map[string]string)
+	vol := b.GetNewVolume(volType, contentType, volStorageName, volConfig)
+
+	// Ensure storage volume settings are honored when doing migration.
+	vol.SetHasSource(false)
+	err = b.driver.FillVolumeConfig(vol)
+	if err != nil {
+		return fmt.Errorf("Failed filling volume config: %w", err)
+	}
+
+	// Check if the volume exists in database
+	dbVol, err := VolumeDBGet(b, inst.Project().Name, inst.Name(), volType)
+	if err != nil && !response.IsNotFoundError(err) {
+		return err
+	}
+
+	if dbVol != nil {
+		return fmt.Errorf("Volume for instance %q already exists in database", inst.Name())
+	}
+
+	// Check if the volume exists on storage.
+	volExists, err := b.driver.HasVolume(vol)
+	if err != nil {
+		return err
+	}
+
+	if volExists {
+		return fmt.Errorf("Volume already exists on storage but not in database")
+	}
+
+	revert := revert.New()
+	defer revert.Fail()
+
+	// Validate config and create database entry for new storage volume if not refreshing.
+	// Strip unsupported config keys (in case the export was made from a different type of storage pool).
+	err = VolumeDBCreate(b, inst.Project().Name, inst.Name(), args.Description, volType, false, vol.Config(), inst.CreationDate(), time.Time{}, contentType, false, true)
+	if err != nil {
+		return err
+	}
+
+	revert.Add(func() { _ = VolumeDBDelete(b, inst.Project().Name, inst.Name(), volType) })
+
+	// Generate the effective root device volume for instance.
+	err = b.applyInstanceRootDiskOverrides(inst, &vol)
+	if err != nil {
+		return err
+	}
+
+	// Override args.Name and args.Config to ensure volume is created based on instance.
+	args.Config = vol.Config()
+	args.Name = inst.Name()
+
+	if slices.Contains(args.ConversionOptions, "format") {
+		// When conversion option "format" is enabled, we need to upload the image to a temporary location
+		// before converting it into the desired format. The conversion cannot be done in-place, therefore
+		// the image has to be saved in an intermediate location.
+		conversionID := fmt.Sprintf("conversion_%s_%s", inst.Project().Name, inst.Name())
+		imgUploadPath := filepath.Join(shared.VarPath("backups"), conversionID)
+
+		var wrapper *ioprogress.ProgressTracker
+		if args.TrackProgress {
+			wrapper = migration.ProgressTracker(op, "block_progress", vol.Name())
+		}
+
+		// Setup progress tracker.
+		fromPipe := io.ReadCloser(conn)
+		if wrapper != nil {
+			fromPipe = &ioprogress.ProgressReader{
+				ReadCloser: fromPipe,
+				Tracker:    wrapper,
+			}
+		}
+
+		// Create new file in backups directory.
+		toFile, err := os.OpenFile(imgUploadPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0)
```

Comment:
What permissions will this file have, as it will be private data we should ensure its only root accessible right? Or does it need to be more relaxed so qemu-img can access it? Can we pass qemu-img a file handle?